Branch-SDK/src/io/branch/indexing/ContentDiscoverer.java

```diff
@@ -0,0 +1,256 @@
+                        contentEvent.put(CONTENT_KEYS_KEY, contentKeysArray);
+
+                        JSONArray contentDataArray = new JSONArray();
+                        contentEvent.put(CONTENT_DATA_KEY, contentDataArray);
+                        discoverFilteredViewContents(filteredElements, contentDataArray, contentKeysArray, activity, isClearText);
+                    } else { // If filter is absent discover all text field keys
+                        if (!discoveredViewList_.contains(viewName)) {  // Check if the view is already discovered. If already discovered no need to get view content keys again
+                            JSONArray contentKeysArray = new JSONArray();
+                            contentEvent.put(CONTENT_KEYS_KEY, contentKeysArray);
+                            discoverViewContents(rootView, null, contentKeysArray, activity.getResources(), isClearText);
+                        }
+                    }
+                    discoveredViewList_.add(viewName);
+
+                    // Cache the analytics data for future use
+                    PrefHelper.getInstance(activity).saveBranchAnalyticsData(contentEvent);
+                    lastActivityReference_ = null;
+                }
+
+            } catch (JSONException ignore) {
+            }
+        }
+    };
+
+
+    private void discoverFilteredViewContents(JSONArray viewIDArray, JSONArray contentDataArray, JSONArray contentKeysArray, Activity activity, boolean isClearText) {
+        try {
+            for (int i = 0; i < viewIDArray.length(); i++) {
+                String viewName = viewIDArray.getString(i);
+                int id = activity.getResources().getIdentifier(viewIDArray.getString(i), "id", activity.getPackageName());
+                View childView = activity.findViewById(id);
+                updateElementData(viewName, childView, isClearText, contentDataArray, contentKeysArray);
+            }
+        } catch (JSONException ignore) {
+
+        }
+    }
+
+    private void discoverViewContents(ViewGroup viewGroup, JSONArray contentDataArray, JSONArray contentKeysArray, Resources res, boolean isClearText) {
+        for (int i = 0; i < viewGroup.getChildCount(); i++) {
+            View childView = viewGroup.getChildAt(i);
+            if (childView.getVisibility() == View.VISIBLE) if (childView instanceof ViewGroup) {
+                discoverViewContents((ViewGroup) childView, contentDataArray, contentKeysArray, res, isClearText);
+            } else {
+                String viewName = res.getResourceEntryName(childView.getId());
+                updateElementData(viewName, childView, isClearText, contentDataArray, contentKeysArray);
+            }
+        }
+    }
+
+    private void updateElementData(String viewName, View view, boolean isClearText, JSONArray contentDataArray, JSONArray contentKeysArray) {
+        String viewVal;
+        if (view instanceof TextView) {
```

Comment:
not sure if this handles nulls, but we should definitely do that in the case that our server-saved ID doesn't return anything (referring to line 173)